### ISSUE TITLE
Kafka sink silently discards events on connection errors

### ISSUE THREAD

**Issue by frankh** [vectordotdev/vector#21031]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

I've noticed the `vector_component_discarded_events_total` metric shows some events are discarded from our kafka sink, despite the fact we have acknowledgements enabled

From correlating logs when the discards happen it looks like this happens every time there is an intermittent connection failure with Kafka

Silent errors on acknowledged sinks is unacceptable and is a complete blocker for us to use Vector sadly. Is there any way to have the sink NACK these events?

### Configuration

```text
kafka:
  type: kafka
    acknowledgements:
      enabled: true
    buffer:
      - type: memory
        max_events: 10000
        when_full: block
```

### Version

0.40.0

### References

Related: 
- https://github.com/vectordotdev/vector/issues/12217
- https://github.com/vectordotdev/vector/issues/10870

**Comment by jszwedko** [comment]
Thanks for this report @frankh ! It does sound like there might be a bug with nacking in the sink for connection issues. Can you share:

- The full config (or a minimal config that is capable of reproducing the issue)
- The log output you get

**Comment by frankh** [comment]
I've made a fix to the Kafka sink so these will get correctly reported as 500 errors: https://github.com/vectordotdev/vector/pull/21036

I'm not 100% sure on when exactly Rejected vs Errored should be sent, but based on the HTTP source's response codes I assume Rejected should mean the event itself is bad, and Errored means the sink failed for reasons unrelated to the event content

**Comment by frankh** [comment]
fixed in #21036
